include/zephyr/net/mqtt_sn.h

```diff
@@ -0,0 +1,290 @@
+ *        application.
+ *
+ * @param[in] client Identifies the client for which the event is notified.
+ * @param[in] evt Event description along with result and associated
+ *                parameters (if any).
+ */
+typedef void (*mqtt_sn_evt_cb_t)(struct mqtt_sn_client *client, const struct mqtt_sn_evt *evt);
+
+/**
+ * @brief Structure to describe an MQTT-SN transport.
+ *
+ * MQTT-SN does not require transports to be reliable or to hold a connection.
+ * Transports just need to be frame-based, so you can use UDP, ZigBee, or even
+ * a simple UART, given some kind of framing protocol is used.
+ */
+struct mqtt_sn_transport {
+	/**
+	 * @brief Will be called once on client init to initialize the transport.
+	 *
+	 * Use this to bind sockets or similar.
+	 */
+	int (*init)(struct mqtt_sn_transport *transport);
+
+	/** Will be called by the library when it wants to send a message. */
+	int (*msg_send)(struct mqtt_sn_client *client, struct mqtt_sn_msg *msg,
+			const void *recipient);
+
+	/** Will be called by the library when it wants to broadcast a message. */
+	int (*msg_broadcast)(struct mqtt_sn_client *client, struct mqtt_sn_msg *msg,
+			     uint8_t radius);
+
+	/**
+	 * @brief Will be called by the library when it wants to receive a message.
+	 *
+	 * Implementations should follow recv conventions.
+	 */
+	ssize_t (*recv)(struct mqtt_sn_client *client, void *buffer, size_t length);
+
+	/**
+	 * @brief Check if incoming data is available.
+	 *
+	 * If poll() returns a positive number, recv must not block.
+	 *
+	 * @return ssize_t number of bytes in the next frame or zero if there is none.
+	 * Negative values signal errors.
+	 */
+	ssize_t (*poll)(struct mqtt_sn_client *client);
+};
+
+#ifdef CONFIG_MQTT_SN_TRANSPORT_UDP
+struct mqtt_sn_transport_udp {
+	struct mqtt_sn_transport tp;
+
+	int sock;
+
+	sockaddr_in gwaddr;
+	socklen_t gwaddrlen;
+};
+
+#define UDP_TRANSPORT(transport) CONTAINER_OF(transport, struct mqtt_sn_transport, tp)
+
+void mqtt_sn_transport_udp_init(struct mqtt_sn_transport_udp *udp);
+#endif
+
+/**
+ * Structure describing an MQTT-SN client.
+ */
+struct mqtt_sn_client {
+	struct mqtt_sn_data client_id; /**< 1-23 character unique client ID */
+
+	struct mqtt_sn_data will_topic; /**< Must be initialized before connecting with will=true */
+	struct mqtt_sn_data will_msg; /**< Must be initialized before connecting with will=true */
+	enum mqtt_sn_qos will_qos;
+	bool will_retain;
+
+	struct mqtt_sn_transport *transport;
+
+	mqtt_sn_evt_cb_t evt_cb;
+
+	uint16_t next_msg_id;
+	sys_slist_t publish;
+	sys_slist_t topic;
+
+	int state;
+
+	struct k_work_delayable wrk;
+};
+
+/**
+ * Initialize a client.
+ */
+int mqtt_sn_client_init(struct mqtt_sn_client *client, const struct mqtt_sn_data *client_id,
+			struct mqtt_sn_transport *transport, mqtt_sn_evt_cb_t evt_cb);
+
+#ifdef CONFIG_MQTT_SN_TRANSPORT_UDP
+int mqtt_sn_client_init_udp(struct mqtt_sn_client *client, const struct mqtt_sn_data *client_id,
```

Comment:
Is this function actually implemented anywhere?

Reply:
no, removed this